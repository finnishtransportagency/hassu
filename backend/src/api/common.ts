import {
  LataaProjektiQueryVariables,
  ListaaVelhoProjektitQueryVariables,
  TallennaProjektiInput,
  TilaSiirtymaInput,
} from "../../../common/graphql/apiModel";
import { AppSyncResolverEvent } from "aws-lambda/trigger/appsync-resolver";
import { getCorrelationId, setupLambdaMonitoring, setupLambdaMonitoringMetaData, wrapXRayAsync } from "../aws/monitoring";
import { log, setLogContextOid } from "../logger";
import { identifyUser } from "../user";
import { ClientError } from "../error/ClientError";
import { SystemError } from "../error/SystemError";
import { IllegalAineistoStateError } from "../error/IllegalAineistoStateError";
import { NotFoundError } from "../error/NotFoundError";
import { IllegalAccessError } from "../error/IllegalAccessError";

export type AppSyncEventArguments =
  | unknown
  | LataaProjektiQueryVariables
  | ListaaVelhoProjektitQueryVariables
  | TallennaProjektiInput
  | TilaSiirtymaInput;

export type LambdaResult = {
  data: unknown;
  correlationId: string;
};

export async function commonHandleEvent(
  event: AppSyncResolverEvent<AppSyncEventArguments>,
  executeOperation: (event: AppSyncResolverEvent<AppSyncEventArguments>) => Promise<unknown>
): Promise<LambdaResult> {
  setupLambdaMonitoring();
  if (log.isLevelEnabled("debug")) {
    log.debug({ event });
  }

  return wrapXRayAsync("handler", async (subsegment) => {
    setupLambdaMonitoringMetaData(subsegment);

    try {
      await identifyUser(event);
      const data = await executeOperation(event);
      const corId = getCorrelationId();
      const lambdaResult: LambdaResult = { data, correlationId: corId || "" };
      return lambdaResult;
    } catch (e: unknown) {
      if (e instanceof Error) {
        // Only data that is sent out in case of error is the error message. We wish to log correlationId with the
        // error, so the only way to do it is to encode the data into error message field. The error field is decoded
        // in deployment/lib/template/response.vtl

        let errorType = "Error";
        let errorSubType = "(no subtype)";
        if (
          e instanceof NotFoundError ||
          e instanceof IllegalAineistoStateError ||
          e instanceof IllegalAccessError ||
          e instanceof ClientError
        ) {
          errorType = "ClientError";
          errorSubType = e.className;
          log.error(e.message);
        } else if (e instanceof SystemError) {
          errorType = "SystemError";
          errorSubType = e.className;
          log.error(e);
        } else {
          log.error(e);
        }

        e.message = JSON.stringify({
          message: e.message,
          correlationId: getCorrelationId(),
          errorType,
          errorSubType,
        });
      }
      throw e;
    } finally {
      setLogContextOid(undefined);
    }
  });
}
