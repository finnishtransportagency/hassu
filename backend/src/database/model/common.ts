import {
  AineistoTila,
  ELY,
  EVK,
  IlmoitettavaViranomainen,
  Kieli,
  ProjektiTyyppi,
  SuunnittelustaVastaavaViranomainen,
  LadattuTiedostoTila,
} from "hassu-common/graphql/apiModel";

export type LocalizedMap<T> = { [key in Kieli]?: T } | null;
export type RequiredLocalizedMap<T> = {
  [Kieli.SUOMI]: T;
  [Kieli.RUOTSI]?: T;
};

export enum SaameKieli {
  POHJOISSAAME = "POHJOISSAAME",
}

export enum SuomiRuotsiKieli {
  SUOMI = "SUOMI",
  RUOTSI = "RUOTSI",
}

export type SaameLocalizedMap<T> = { [key in SaameKieli]?: T } | null;

export type Kielitiedot = {
  ensisijainenKieli: Kieli;
  toissijainenKieli?: Kieli | null;
  projektinNimiVieraskielella?: string;
};

export type Yhteystieto = {
  etunimi: string;
  sukunimi: string;
  organisaatio?: string;
  kunta?: number | null;
  elyOrganisaatio?: ELY;
  evkOrganisaatio?: EVK;
  puhelinnumero: string;
  sahkoposti: string;
  titteli?: string;
};

export type Aineisto = {
  // Dokumentin oid Velhossa
  dokumenttiOid: string;
  // Kategorian ID, joka viittaa kategoriapuun kategoriaan
  kategoriaId?: string;
  // Suhteellinen polku tiedostoon yllapidon S3-bucketissa projektin alla
  tiedosto?: string;
  // Tiedostonimi naytettavaksi
  nimi: string;
  // Tunniste, joka on luotu FE:ssä, kun tiedosto on valittu. Ei muutu.
  uuid: string;
  // Jos tosi, tallennuksen validoinnissa ja tallennuksessa käytetään tunnisteena uuid:n sijaa dokumenttiOid:ta
  // HUOM! tätä tietoa ei ole oikeasti tarkoitus tallentaa kantaan
  uuidGeneratedBySchemaMigration?: boolean;
  // Aikaleima, milloin tiedosto on tuotu jarjestelmaan yyyy-MM-ddTHH:mm
  tuotu?: string;
  // Numero jarjestamista varten
  jarjestys?: number | null;
  tila: AineistoTila;
  kategoriaMuuttunut?: boolean;
};

export type AineistoNew = ILadattuTiedosto & {
  // Dokumentin oid Velhossa
  dokumenttiOid: string;
  // Kategorian ID, joka viittaa kategoriapuun kategoriaan
  kategoriaId?: string;
};

export type StandardiYhteystiedot = {
  yhteysTiedot?: Yhteystieto[];
  yhteysHenkilot?: string[];
};

export type IlmoituksenVastaanottajat = {
  kunnat?: Array<KuntaVastaanottaja> | null;
  viranomaiset?: Array<ViranomaisVastaanottaja> | null;
  maakunnat?: Array<MaakuntaVastaanottaja> | null;
};

export type SahkopostiVastaanottaja = {
  sahkoposti: string;
  messageId?: string;
  lahetetty?: string | null;
  lahetysvirhe?: boolean | null;
};

export type KuntaVastaanottaja = {
  id: number;
} & SahkopostiVastaanottaja;

export type MaakuntaVastaanottaja = {
  id: number;
} & SahkopostiVastaanottaja;

export type ViranomaisVastaanottaja = {
  nimi: IlmoitettavaViranomainen;
} & SahkopostiVastaanottaja;

export type UudelleenKuulutus = {
  tila: UudelleenkuulutusTila;
  selosteKuulutukselle?: RequiredLocalizedMap<string>;
  selosteLahetekirjeeseen?: RequiredLocalizedMap<string>;
  alkuperainenHyvaksymisPaiva?: string;
  tiedotaKiinteistonomistajia?: boolean;
  alkuperainenKuulutusPaiva?: string;
};

export type AineistoMuokkaus = {
  alkuperainenHyvaksymisPaiva?: string;
};

export enum UudelleenkuulutusTila {
  PERUUTETTU = "PERUUTETTU",
  JULKAISTU_PERUUTETTU = "JULKAISTU_PERUUTETTU",
}

export type LinkitettyVelhoProjekti = {
  oid: string;
  nimi: string;
  tyyppi?: ProjektiTyyppi | null;
};

export type Velho = {
  nimi: string;
  tyyppi?: ProjektiTyyppi | null;
  kuvaus?: string | null;
  vaylamuoto?: string[] | null;
  asiatunnusVayla?: string | null;
  asiatunnusELY?: string | null;
  asiatunnusEVK?: string | null;
  suunnittelustaVastaavaViranomainen?: SuunnittelustaVastaavaViranomainen | null;
  toteuttavaOrganisaatio?: string | null;
  vastuuhenkilonNimi?: string | null;
  vastuuhenkilonEmail?: string | null;
  varahenkilonNimi?: string | null;
  varahenkilonEmail?: string | null;
  maakunnat?: number[] | null;
  kunnat?: number[] | null;
  linkki?: string | null;
  linkitetytProjektit?: LinkitettyVelhoProjekti[] | null;
  geoJSON?: string | null;
};
export interface ILadattuTiedosto {
  // Kayttajalle esitettava tiedostonimi
  nimi: string;
  uuid: string;
  lisatty: string;
}

export type LadattuTiedostoNew = ILadattuTiedosto;

export type LadattuTiedosto = {
  // Suhteellinen polku tiedostoon yllapidon S3-bucketissa projektin alla
  tiedosto: string;
  // Kayttajalle esitettava tiedostonimi
  nimi?: string | null;
  uuid: string;
  tuotu?: string | null;
  //Numero jarjestamista varten
  jarjestys?: number;
  tila: LadattuTiedostoTila;
};

export type KunnallinenLadattuTiedosto = {
  kunta: number;
} & ILadattuTiedosto;

export type Laskutustiedot = {
  ovtTunnus?: string | null;
  verkkolaskuoperaattorinTunnus?: string | null;
  viitetieto?: string | null;
};
