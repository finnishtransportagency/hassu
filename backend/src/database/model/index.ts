export * from "./common";
export * from "./projekti";
export * from "./suunnitteluVaihe";
export * from "./nahtavillaoloVaihe";
export * from "./hyvaksymisPaatosVaihe";
