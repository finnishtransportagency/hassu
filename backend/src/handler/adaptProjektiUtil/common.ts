import * as API from "../../../../common/graphql/apiModel";

export function adaptIlmoituksenVastaanottajat(
  vastaanottajat: API.IlmoituksenVastaanottajat | null | undefined
): API.IlmoituksenVastaanottajat {
  if (!vastaanottajat) {
    return vastaanottajat as null | undefined;
  }
  const kunnat: API.KuntaVastaanottaja[] =
    vastaanottajat?.kunnat?.map((kunta) => ({ __typename: "KuntaVastaanottaja", ...kunta })) || null;
  const viranomaiset: API.ViranomaisVastaanottaja[] =
    vastaanottajat?.viranomaiset?.map((viranomainen) => ({
      __typename: "ViranomaisVastaanottaja",
      ...viranomainen,
    })) || null;
  return { __typename: "IlmoituksenVastaanottajat", kunnat, viranomaiset };
}
