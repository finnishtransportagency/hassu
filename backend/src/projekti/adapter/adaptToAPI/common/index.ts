export * from "./lisaaTypename";
export * from "./adaptLadattuTiedostoToAPI";
export * from "./adaptKuulutusSaamePDFtToAPI";
export * from "./adaptLokalisoituTekstiToAPI";
export * from "./adaptPakotettuLokalisoituTekstiToAPI";
export * from "./adaptLogotToAPI";
export * from "./adaptUudelleenKuulutusToAPI";
export * from "./adaptAineistoMuokkausToAPI";
export * from "./adaptAineistotToAPI";
export * from "./adaptIlmoituksenVastaanottajatToAPI";
export * from "./adaptLokalisoituLinkkiToAPI";
export * from "./adaptStandardiYhteystiedotToAPIByAddingProjektiPaallikko";
export * from "./adaptVelhoToAPI";
