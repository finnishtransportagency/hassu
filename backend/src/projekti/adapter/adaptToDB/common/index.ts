export * from "./adaptTiedostotToSave";
export * from "./removeTypeName";
export * from "./adaptLogoFilesToSave";
export * from "./adaptIlmoituksenVastaanottajatToSave";
export * from "./adaptYhteystiedotToSave";
export * from "./adaptAineistotToSave";
export * from "./adaptHankkeenKuvausToSave";
export * from "./adaptLokalisoituTekstiToSave";
export * from "./adaptLokalisoituLinkkiToSave";
export * from "./getId";
export * from "./forEverySaameDoAsync";
export * from "./adaptLadattuTiedostoToSave";
