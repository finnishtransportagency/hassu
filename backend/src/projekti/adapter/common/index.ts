export * from "./adaptAineistot";
export * from "./adaptIlmoituksenVastaanottajat";
export * from "./util";
export * from "./lisaaTypename";
export * from "./adaptHankkeenKuvaus";
