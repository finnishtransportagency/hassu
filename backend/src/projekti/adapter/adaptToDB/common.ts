import * as API from "../../../../../common/graphql/apiModel";
import { IllegalArgumentError } from "../../../error/IllegalArgumentError";
import { Aineisto, LocalizedMap } from "../../../database/model";
import { AineistoChangedEvent, ProjektiAdaptationResult, ProjektiEventType } from "../projektiAdapter";
import remove from "lodash/remove";

export function adaptIlmoituksenVastaanottajatToSave(
  vastaanottajat: API.IlmoituksenVastaanottajatInput | null | undefined
): API.IlmoituksenVastaanottajat {
  if (!vastaanottajat) {
    return vastaanottajat as null | undefined;
  }
  const kunnat: API.KuntaVastaanottaja[] = vastaanottajat?.kunnat?.map((kunta) => ({ __typename: "KuntaVastaanottaja", ...kunta })) || null;
  if (!vastaanottajat?.viranomaiset || vastaanottajat.viranomaiset.length === 0) {
    throw new IllegalArgumentError("Viranomaisvastaanottajia pitää olla vähintään yksi.");
  }
  const viranomaiset: API.ViranomaisVastaanottaja[] =
    vastaanottajat?.viranomaiset?.map((viranomainen) => ({
      __typename: "ViranomaisVastaanottaja",
      ...viranomainen,
    })) || null;
  return { __typename: "IlmoituksenVastaanottajat", kunnat, viranomaiset };
}

export function adaptYhteystiedotToSave(yhteystietoInputs: Array<API.YhteystietoInput>): API.YhteystietoInput[] | undefined {
  return yhteystietoInputs?.length > 0 ? yhteystietoInputs.map((yt) => ({ ...yt })) : undefined;
}

export function adaptYhteysHenkilotToSave(yhteystiedot: string[]): string[] {
  return yhteystiedot.filter((yt, index) => yhteystiedot.indexOf(yt) === index);
}

export function adaptStandardiYhteystiedot(kuulutusYhteystiedot: API.StandardiYhteystiedotInput): API.StandardiYhteystiedotInput {
  return {
    yhteysTiedot: adaptYhteystiedotToSave(kuulutusYhteystiedot.yhteysTiedot),
    yhteysHenkilot: adaptYhteysHenkilotToSave(kuulutusYhteystiedot.yhteysHenkilot),
  };
}

export function adaptHankkeenKuvausToSave(hankkeenKuvaus: API.HankkeenKuvauksetInput): LocalizedMap<string> {
  if (!hankkeenKuvaus) {
    return undefined;
  }
  return { ...hankkeenKuvaus };
}

export function adaptAineistotToSave(
  dbAineistot: Aineisto[] | undefined,
  aineistotInput: API.AineistoInput[] | undefined,
  projektiAdaptationResult: Partial<ProjektiAdaptationResult>
): Aineisto[] | undefined {
  const resultAineistot = [];
  let hasPendingChanges = undefined;

  // Examine and update existing documents
  if (dbAineistot) {
    dbAineistot.forEach((dbAineisto) => {
      const updateAineistoInput = pickAineistoFromInputByDocumenttiOid(aineistotInput, dbAineisto.dokumenttiOid);
      if (updateAineistoInput) {
        // Update existing one

        if (dbAineisto.nimi !== updateAineistoInput.nimi) {
          hasPendingChanges = true;
        }
        dbAineisto.nimi = updateAineistoInput.nimi;
        dbAineisto.jarjestys = updateAineistoInput.jarjestys;
        dbAineisto.kategoriaId = updateAineistoInput.kategoriaId;
        resultAineistot.push(dbAineisto);
      }
      if (!updateAineistoInput && aineistotInput) {
        dbAineisto.tila = API.AineistoTila.ODOTTAA_POISTOA;
        resultAineistot.push(dbAineisto);
        hasPendingChanges = true;
      }
    });
  }

  // Add new ones and optionally trigger import later
  if (aineistotInput) {
    for (const aineistoInput of aineistotInput) {
      resultAineistot.push({
        dokumenttiOid: aineistoInput.dokumenttiOid,
        nimi: aineistoInput.nimi,
        kategoriaId: aineistoInput.kategoriaId,
        jarjestys: aineistoInput.jarjestys,
        tila: API.AineistoTila.ODOTTAA_TUONTIA,
      });
      hasPendingChanges = true;
    }
  }

  if (hasPendingChanges) {
    projektiAdaptationResult.pushEvent({ eventType: ProjektiEventType.AINEISTO_CHANGED } as AineistoChangedEvent);
  }
  return resultAineistot;
}

function pickAineistoFromInputByDocumenttiOid(aineistotInput: API.AineistoInput[], dokumenttiOid: string) {
  const matchedElements = remove(aineistotInput, { dokumenttiOid });
  if (matchedElements.length > 0) {
    return matchedElements[0];
  }
  return undefined;
}
