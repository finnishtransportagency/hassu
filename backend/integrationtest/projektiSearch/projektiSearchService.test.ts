import { describe, it } from "mocha";
import { ProjektiFixture } from "../../test/fixture/projektiFixture";
import { projektiSearchService } from "../../src/projektiSearch/projektiSearchService";
import { Kieli, ListaaProjektitInput, ProjektiTyyppi, Status, SuunnittelustaVastaavaViranomainen } from "../../../common/graphql/apiModel";
import { DBProjekti } from "../../src/database/model";
import dayjs from "dayjs";
import { UserFixture } from "../../test/fixture/userFixture";
import { userService } from "../../src/user";
import sinon from "sinon";
import assert from "assert";
import { kuntametadata } from "../../../common/kuntametadata";

const { expect } = require("chai");

const delay = (ms: number) => new Promise((res) => setTimeout(res, ms));

type ProjektiVariation = {
  oid: string;
  suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen;
  maakunnat: number[];
  kunnat: number[];
} & Omit<ListaaProjektitInput, "vaihe" | "suunnittelustaVastaavaViranomainen">;

const testData: ProjektiVariation[] = [
  {
    oid: "1",
    projektiTyyppi: ProjektiTyyppi.TIE,
    vaylamuoto: ["tie"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Pohjanmaa"]),
    kunnat: kuntametadata.idsForKuntaNames(["Tampere", "Nokia"]),
    asiatunnus: "A1",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.POHJOIS_POHJANMAAN_ELY,
  },
  {
    oid: "2",
    projektiTyyppi: ProjektiTyyppi.TIE,
    vaylamuoto: ["tie"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Uusimaa", "Pirkanmaa"]),
    kunnat: kuntametadata.idsForKuntaNames(["Tampere", "Helsinki"]),
    asiatunnus: "A2",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.UUDENMAAN_ELY,
  },
  {
    oid: "3",
    projektiTyyppi: ProjektiTyyppi.RATA,
    vaylamuoto: ["rata"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Ahvenanmaa"]),
    kunnat: kuntametadata.idsForKuntaNames(["Maarianhamina"]),
    asiatunnus: "A3",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.VAYLAVIRASTO,
  },
  {
    oid: "4",
    projektiTyyppi: ProjektiTyyppi.RATA,
    vaylamuoto: ["tie", "rata"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Uusimaa", "Ahvenanmaa"]),
    kunnat: kuntametadata.idsForKuntaNames(["Maarianhamina", "Helsinki"]),
    asiatunnus: "A4",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.VAYLAVIRASTO,
  },
  {
    oid: "5",
    projektiTyyppi: ProjektiTyyppi.YLEINEN,
    vaylamuoto: ["tie"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Pirkanmaa"]),
    kunnat: kuntametadata.idsForKuntaNames(["Tampere"]),
    asiatunnus: "A5",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.PIRKANMAAN_ELY,
  },
  {
    oid: "6",
    projektiTyyppi: ProjektiTyyppi.YLEINEN,
    vaylamuoto: ["tie", "rata"],
    maakunnat: kuntametadata.idsForMaakuntaNames(["Kanta-Häme"]),
    kunnat: kuntametadata.idsForKuntaNames(["Hämeenlinna"]),
    asiatunnus: "A6",
    suunnittelustaVastaavaViranomainen: SuunnittelustaVastaavaViranomainen.KESKI_SUOMEN_ELY,
  },
];

describe.skip("ProjektiSearchService", () => {
  const projektiFixture = new ProjektiFixture();
  let userFixture: UserFixture;

  before(async () => {
    for (let i = 0; i < testData.length; i++) {
      const data = testData[i];
      const projekti: DBProjekti = projektiFixture.dbProjekti2();
      assert(projekti);
      assert(projekti.velho);
      assert(projekti.aloitusKuulutusJulkaisut);
      projekti.oid = data.oid;
      projekti.velho.tyyppi = data.projektiTyyppi;
      projekti.velho.vaylamuoto = data.vaylamuoto;
      projekti.velho.nimi = "unittest" + data.oid;
      projekti.velho.kunnat = data.kunnat;
      projekti.velho.maakunnat = data.maakunnat;
      projekti.velho.asiatunnusELY = data.asiatunnus;
      projekti.suunnittelustaVastaavaViranomainen = data.suunnittelustaVastaavaViranomainen;
      projekti.paivitetty = dayjs("2020-01-01T23:00:00+02:00").add(-i, "hours").format();
      projekti.aloitusKuulutusJulkaisut[0].kuulutusPaiva = "2000-02-02";
      projekti.aloitusKuulutusJulkaisut[0].siirtyySuunnitteluVaiheeseen = "2222-02-02";
      await projektiSearchService.indexProjekti(projekti);
    }
    await delay(500);
  });

  // after(async () => {
  //   for (const data of testData) {
  //     await projektiSearchService.removeProjekti(data.oid);
  //   }
  // });

  beforeEach(() => {
    userFixture = new UserFixture(userService);
  });

  afterEach(() => {
    userFixture.logout();
    sinon.restore();
  });

  it("should search by list of oids successfully", async () => {
    const dbProjektis = await projektiSearchService.searchByOid(["1", "2"]);
    expect(dbProjektis).toMatchSnapshot();
  });

  it("should search by name successfully", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        nimi: "unittest3",
        projektiTyyppi: ProjektiTyyppi.RATA,
      })
    ).tulokset;

    expect(results);
    expect(results?.map((result) => result.oid)).to.contain("3");
  });

  it("should search by vaihe successfully", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        vaihe: [Status.SUUNNITTELU],
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.vaihe).to.eq(Status.SUUNNITTELU, JSON.stringify(result));
    }
  });

  it("should search by suunnittelustaVastaavaViranomainen successfully", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        suunnittelustaVastaavaViranomainen: [
          SuunnittelustaVastaavaViranomainen.KESKI_SUOMEN_ELY,
          SuunnittelustaVastaavaViranomainen.UUDENMAAN_ELY,
        ],
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.suunnittelustaVastaavaViranomainen).to.be.oneOf([
        SuunnittelustaVastaavaViranomainen.KESKI_SUOMEN_ELY,
        SuunnittelustaVastaavaViranomainen.UUDENMAAN_ELY,
      ]);
    }
  });

  it("should search by vaihe and name successfully", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        nimi: "unittest3",
        projektiTyyppi: ProjektiTyyppi.RATA,
        vaihe: [Status.SUUNNITTELU],
      })
    ).tulokset;
    assert(results);
    expect(results.map((result) => result.oid)).to.contain("3");
  });

  it("should not include results if name and vaihe don't match at the same time", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        nimi: "unittest3",
        projektiTyyppi: ProjektiTyyppi.RATA,
        vaihe: [Status.EI_JULKAISTU],
      })
    ).tulokset;
    expect(results).to.have.length(0);
  });

  it("should search by asiatunnus", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        asiatunnus: "A3",
        projektiTyyppi: ProjektiTyyppi.RATA,
      })
    ).tulokset;
    expect(results).to.have.length(1);
    assert(results);
    expect(results[0].oid).to.eq("3");
  });

  it("should search by vaylamuoto", async () => {
    const results = (
      await projektiSearchService.searchYllapito({
        vaylamuoto: ["tie"],
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.vaylamuoto).to.contain.oneOf(["tie"]);
    }
  });

  let uusimaaPirkanmaa = kuntametadata.idsForMaakuntaNames(["Uusimaa", "Pirkanmaa"]);
  it("should search by maakunta", async () => {
    userFixture.loginAs(UserFixture.mattiMeikalainen);
    const results = (
      await projektiSearchService.searchYllapito({
        maakunta: uusimaaPirkanmaa,
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.maakunnat).to.contain.oneOf(["Uusimaa", "Pirkanmaa"]);
    }
  });

  it("should search by modification permissions", async () => {
    userFixture.loginAs(UserFixture.mattiMeikalainen);
    const results = (
      await projektiSearchService.searchYllapito({
        vainProjektitMuokkausOikeuksin: true,
      })
    ).tulokset;
    assert(results);
    expect(results.map((value) => value.oid)).to.contain.members(["1", "2"]);

    userFixture.loginAs(UserFixture.manuMuokkaaja);
    const results2 = (
      await projektiSearchService.searchYllapito({
        vainProjektitMuokkausOikeuksin: true,
      })
    ).tulokset;
    expect(results2).to.have.length(0);
  });

  it("should search by maakunta as kansalainen", async () => {
    const results = (
      await projektiSearchService.searchJulkinen({
        kieli: Kieli.SUOMI,
        maakunta: uusimaaPirkanmaa,
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.maakunnat).to.contain.oneOf(uusimaaPirkanmaa);
    }
  });

  it("should search by kunta as kansalainen", async () => {
    let maarianhamina = kuntametadata.idForKuntaName("Maarianhamina");
    const results = (
      await projektiSearchService.searchJulkinen({
        kieli: Kieli.SUOMI,
        kunta: [maarianhamina],
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.kunnat).to.contain(maarianhamina);
    }
  });

  it("should search by nimi as kansalainen", async () => {
    const results = (
      await projektiSearchService.searchJulkinen({
        kieli: Kieli.SUOMI,
        nimi: "testiprojekti",
        vaihe: [Status.ALOITUSKUULUTUS],
      })
    ).tulokset;
    assert(results);
    for (const result of results) {
      expect(result.nimi).to.contain("testiprojekti");
    }
  });
});
