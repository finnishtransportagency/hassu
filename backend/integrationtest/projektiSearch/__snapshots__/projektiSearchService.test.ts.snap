// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjektiSearchService should index, update, search, and delete document to search engine successfully 1`] = `
Object {
  "liittyvatSuunnitelmat": Array [
    Object {
      "asiatunnus": "atunnus123",
      "nimi": "Littyva suunnitelma 1 nimi",
    },
  ],
  "muistiinpano": "Testiprojekti 1:n muistiinpano",
  "oid": "ProjektiSearchService1",
  "velho": Object {
    "kunnat": Array [
      "Tampere",
      "Nokia",
    ],
    "nimi": "Testiprojekti 1",
    "tilaajaOrganisaatio": "Uudenmaan ELY-keskus",
    "tyyppi": "TIE",
    "vaylamuoto": Array [
      "tie",
    ],
  },
}
`;
